### PR TITLE
neofs-node: Reload config for pprof and metrics on SIGHUP

### LINKED ISSUE #1868: Reload pprof/metrics services on SIGHUP

### PR DESCRIPTION
**Pull request by acid-ant**
Close #1868

Signed-off-by: Anton Nikiforov <an.nikiforov@yadro.com>

### CODE REVIEW

**Review by carpawell — COMMENTED** [review]
is it possible to provide some sort of a [dynamic configuration](https://github.com/nspcc-dev/neofs-node/blob/51e886dd67597bc78e420f2324bd1ec9085e4a7e/cmd/neofs-node/config.go#L368-L372) to our http servers the same way it is [done](https://github.com/nspcc-dev/neofs-node/blob/51e886dd67597bc78e420f2324bd1ec9085e4a7e/pkg/util/logger/logger.go#L48-L51) for `logger`? at least that is how i planned SIGHUP unification

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3751

File: cmd/neofs-node/config.go [hunk 1]
```diff
@@ -306,9 +306,11 @@ type internals struct {
 
 	log *logger.Logger
 
-	wg      *sync.WaitGroup
-	workers []worker
-	closers []func()
+	wg           *sync.WaitGroup
+	workers      []*worker
+	namedWorkers map[string]*worker
```
Comment (fyrchik) [inline cmd/neofs-node/config.go:311]: What about storing name in a `worker` and sticking with the old approach here? To me map is an overkill here.

File: cmd/neofs-node/pprof.go [hunk 2]
```diff
@@ -5,42 +5,49 @@ import (
...
+	// Shutdown server
+	if closer, ok := c.namedClosers[profilerServiceKey]; ok {
+		closer()
+	}
+	// Cleanup
+	delete(c.namedClosers, profilerServiceKey)
+	delete(c.namedWorkers, profilerServiceKey)
+	// Init server with new parameters
+	initProfiler(c)
+	//Start worker
```
Comment (fyrchik) [inline cmd/neofs-node/pprof.go:49]: Space after `//` (`gofmt` should fix it)

File: cmd/neofs-node/pprof.go [hunk 3]
```diff
@@ -5,42 +5,49 @@ import (
...
-		err := srv.Shutdown()
-		if err != nil {
-			c.log.Debug("could not shutdown pprof server",
-				zap.String("error", err.Error()),
-			)
-		}
-
-		c.log.Debug("profiling service has been stopped")
-	})
+func reloadProfiler(c *cfg) {
```
Comment (fyrchik) [inline cmd/neofs-node/pprof.go:39]: Can we somehow unify it with `reloadMetrics`? Or at least reusing `shutdown` code?

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3766

File: cmd/neofs-node/config.go (hunk 1 above)
Comment (acid-ant) [reply cmd/neofs-node/config.go:311]: In this case removing from slice will be a bit complex. I see that number of workers is too low to think about this as about time consuming task. Working on implementing of the same approach as was done for logger.

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3767

File: cmd/neofs-node/config.go (hunk 1 above)
Comment (acid-ant) [reply cmd/neofs-node/config.go]: Refactored, please review.

File: cmd/neofs-node/pprof.go (hunk 3 above)
Comment (acid-ant) [reply cmd/neofs-node/pprof.go]: Refactored, please review.

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3768

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3769

File: cmd/neofs-node/httpcmp.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	})
+}
+
+func (cmp *httpCmp) Name() string {
+	return cmp.name
+}
+
+func (cmp *httpCmp) Reload() error {
+	// Shutdown server
+	closer := getCloser(cmp.cfg, cmp.name)
```
[the remark this answers is not in the document]
Comment (acid-ant) [reply cmd/neofs-node/httpcmp.go:59]: Updated.

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3772

File: cmd/neofs-node/httpcmp.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (acid-ant) [reply cmd/neofs-node/httpcmp.go:59]: Fixed.

File: cmd/neofs-node/closer.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+		}
+	}
+	return nil
+}
+
+func delCloser(c *cfg, name string) {
+	for i, clsr := range c.closers {
+		if clsr.name == name {
+			c.closers[i] = c.closers[len(c.closers)-1]
+			c.closers[len(c.closers)-1] = nil
```
Comment (cthulhu-rider) [inline cmd/neofs-node/closer.go:21]: Do we really need this according to next line?

File: cmd/neofs-node/config.go [hunk 6]
```diff
@@ -307,8 +307,8 @@ type internals struct {
 	log *logger.Logger
 
 	wg      *sync.WaitGroup
-	workers []worker
-	closers []func()
+	workers []*worker
```
Comment (cthulhu-rider) [inline cmd/neofs-node/config.go:310]: Why do we need to use slice of pointers? It almost always unnecessary. Same for `closers`

File: cmd/neofs-node/httpcmp.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	"net/http"
+	"time"
+
+	metricsconfig "github.com/nspcc-dev/neofs-node/cmd/neofs-node/config/metrics"
+	profilerconfig "github.com/nspcc-dev/neofs-node/cmd/neofs-node/config/profiler"
+	httputil "github.com/nspcc-dev/neofs-node/pkg/util/http"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type httpCmp struct {
```
Comment (cthulhu-rider) [inline cmd/neofs-node/httpcmp.go:15]: `Cmp` usually stands for compare, lets clarify the name?

File: pkg/util/logger/logger.go [hunk 8]
```diff
@@ -57,6 +57,11 @@ func (p Prm) Reload() error {
 	return p._log.reload(p)
 }
 
+// Name returns component name.
+func (p Prm) Name() string {
```
Comment (cthulhu-rider) [inline pkg/util/logger/logger.go:61]: This is an applicaiton logic, lets implement this on some adapter in `neofs-node`? Current package is for general purpose.

File: cmd/neofs-node/httpcmp.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	address     string
+	name        string
+	handler     http.Handler
+	shutdownDur time.Duration
+	enabled     bool
+	cfg         *cfg
+}
+
+func (cmp *httpCmp) initComponent(c *cfg) {
```
Comment (cthulhu-rider) [inline cmd/neofs-node/httpcmp.go:24]:
```suggestion
func (cmp *httpCmp) init(c *cfg) {
```

File: cmd/neofs-node/closer.go (hunk 5 above)
Comment (acid-ant) [reply cmd/neofs-node/closer.go:21]: Not sure how it should be in golang, the idea is to facilitate gc work.

File: cmd/neofs-node/config.go (hunk 6 above)
Comment (acid-ant) [reply cmd/neofs-node/config.go]: Reverted

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3778

File: pkg/util/logger/logger.go (hunk 8 above)
Comment (acid-ant) [reply pkg/util/logger/logger.go]: Removed.

File: cmd/neofs-node/httpcmp.go (hunk 9 above)
Comment (acid-ant) [reply cmd/neofs-node/httpcmp.go]: Addressed.

**Comment by acid-ant** [comment]
Current behavior for metrics:
- exclude from requests processing when disabling via SIGHUP
- initialized always
- metrics related to node state set always

File: cmd/neofs-node/closer.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+			return &clsr
+		}
+	}
+	return nil
+}
+
+func delCloser(c *cfg, name string) {
+	for i, clsr := range c.closers {
+		if clsr.name == name {
+			c.closers[i] = c.closers[len(c.closers)-1]
```
Comment (carpawell) [inline cmd/neofs-node/closer.go:20]: can't say for sure but think that we can rely on the closers' order now or in the future. not sure

File: cmd/neofs-node/config.go [hunk 11]
```diff
@@ -607,10 +614,8 @@ func initCfg(appCfg *config.Config) *cfg {
 
 	user.IDFromKey(&c.ownerIDFromKey, key.PrivateKey.PublicKey)
 
-	if metricsconfig.Enabled(c.appCfg) {
-		c.metricsCollector = metrics.NewNodeMetrics()
-		netState.metrics = c.metricsCollector
-	}
+	c.metricsCollector = metrics.NewNodeMetrics()
```
Comment (carpawell) [inline cmd/neofs-node/config.go:617]: why did you remove `if`?

File: cmd/neofs-node/config.go [hunk 12]
```diff
@@ -954,7 +968,7 @@ func (c *cfg) configWatcher(ctx context.Context) {
 			}
 
 			for _, component := range components {
-				err = component.cfg.Reload()
+				err = component.reload()
```
Comment (carpawell) [inline cmd/neofs-node/config.go:971]: is it a good idea to call it in such a way? looks like it is a `component`'s `reload` method. do not really think it is bad but anyway
mb call it `reloadFunc`?

on the other hand: why isn't a `Reload` method like for the `logger`?

File: cmd/neofs-node/httpcomponent.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	httputil "github.com/nspcc-dev/neofs-node/pkg/util/http"
+)
+
+type httpComponent struct {
```
Comment (carpawell) [inline cmd/neofs-node/httpcomponent.go:12]: is it possible to move it to `pkg`?

also making it a separate public component could help with parametrizing it like `logger` and implementing `Reload` method i guess

**Comment by carpawell** [comment]
Changelog conflicts.

File: cmd/neofs-node/config.go (hunk 12 above)
Comment (acid-ant) [reply cmd/neofs-node/config.go]: Renamed reload -> reloadFunc

File: cmd/neofs-node/httpcomponent.go (hunk 13 above)
Comment (acid-ant) [reply cmd/neofs-node/httpcomponent.go:12]: I thought about `Reload` when reviewed `neofs-ir` - in ir we also have pprof and metrics. Should we move it to pkg in this PR?

File: cmd/neofs-node/closer.go (hunk 10 above)
Comment (acid-ant) [reply cmd/neofs-node/closer.go:20]: It is a simplest way to detect do we have closer/worker for the httpcomponent or not. In other case we need to remember id for worker and closer and do not forget to set it to negative value to mark that closer/worker for component deleted or not initialized at that moment, and do not forget to check this when reloading. We have less than 20 workers\closers currently. You prefer this approach?

**Review by fyrchik — APPROVED** [review]

File: cmd/neofs-node/worker.go [hunk 14]
```diff
@@ -4,31 +4,47 @@ import (
...
+	go func(w worker) {
+		w.fn(c.ctx)
+		c.wg.Done()
+	}(wrk)
+}
+
+func delWorker(c *cfg, name string) {
+	for i, worker := range c.workers {
+		if worker.name == name {
+			c.workers[i] = c.workers[len(c.workers)-1]
```
[the remark this answers is not in the document]
Comment (acid-ant) [reply cmd/neofs-node/worker.go:36]: The length of the workers is not fixed, we can't keep empty worker/closer to keep order in this array for disabled service.

File: pkg/network/transport/object/grpc/service.go [hunk 15]
```diff
@@ -24,6 +24,11 @@ func New(c objectSvc.ServiceServer) *Server {
 	}
 }
 
+// SetSrv set service server.
+func (s *Server) SetSrv(c objectSvc.ServiceServer) {
```
[the remark this answers is not in the document]
Comment (acid-ant) [reply pkg/network/transport/object/grpc/service.go:28]: Added because we need to exclude/include metrics request handler from chain. When metrics disabled I set for object service as next service in chain sign service, when enabled - metrics. Agree that it is better to disable calculation inside metrics service by flag.

File: pkg/util/http/server.go [hunk 16]
```diff
@@ -85,3 +89,14 @@ func New(prm Prm, opts ...Option) *Server {
 		},
 	}
 }
+
+// NewPrm creates a new instance of the Prm.
+//
+// Panics if at least one value of the parameters is invalid.
+func NewPrm(addr string, handler http.Handler) *Prm {
```
[the remark this answers is not in the document]
Comment (acid-ant) [reply pkg/util/http/server.go]: Renamed - Prm -> HTTPSrvPrm

File: cmd/neofs-node/worker.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply cmd/neofs-node/worker.go:36]: yes, but still, why sth like `c.workers = append(c.workers[:i], c.workers[i+1:]...)` doen't work?

[the remark this answers is not in the document]
Comment (acid-ant) [reply cmd/neofs-node/worker.go]: Oh, you about this, yes your approach is better for reading and keeps order. Updated.
